app/src/main/java/com/duckduckgo/mobile/android/duckduckgo/data/base/JsonEntity.java

```diff
@@ -0,0 +1,17 @@
+package com.duckduckgo.mobile.android.duckduckgo.data.base;
+
+/**
+ * Created by fgei on 6/12/17.
+ */
+
+public abstract class JsonEntity {
```

Comment:
@subsymbolic so something like this?

`public class SharedPreferencesDataStore<T extends JsonEntity> {

    private SharedPreferences sharedPreferences;
    private EntityCreator<T> entityCreator;

    public static abstract class EntityCreator<T> {
        public abstract T newInstance();
    }

    public SharedPreferencesDataStore(Context context, String fileName, EntityCreator<T> entityCreator) {
        sharedPreferences = context.getSharedPreferences(fileName, Context.MODE_PRIVATE);
        this.entityCreator = entityCreator;
    }

    public List<T> getAll() {
        List<T> out = new ArrayList<>();
        for (Object item : sharedPreferences.getAll().values()) {
            T t = entityCreator.newInstance();
            t.fromJson(item.toString());
            out.add(t);

        }
        return out;
    }
 ...
}

public class BookmarkSharedPreferences extends SharedPreferencesDataStore<BookmarkJsonEntity> {

    private static final String PREF_NAME = "bookmarks";

    public BookmarkSharedPreferences(Context context) {
        super(context, PREF_NAME, new EntityCreator<BookmarkJsonEntity>() {
            @Override
            public BookmarkJsonEntity newInstance() {
                return new BookmarkJsonEntity();
            }
        });
    }
}`